### DISCUSSION TITLE
Differencer in ForecastingPipeline and TransformedTargetForecaster does not work!

### DISCUSSION THREAD

**Discussion by anthonygiorgio97** [alan-turing-institute/sktime#2451] [Q&A]
I was trying to do a Pipeline with transformers for both exogenous X (4 features) and endogenous y variable with a FeatureSelection(), a LogTransformer() and a Differencer() combined with a cross validation to tune Prophet (but the same happen with other algorithms). The difference operator must exclude the first index from both exogenous and endogenous variable (because I have specified this transformation both in ForecastingPipeline and TransformedTargetForecaster), but when I call the fit method this error raise: 

(time) indices of series 0 are not contained in index of series 0, extra indices are: DatetimeIndex(['2016-01-31'], dtype='datetime64[ns]', freq=None)

where 2016-01-31 is the first index. If I use a classic Detrender() the code work, but I want to try with the Differencer() operator.

I am using sktime 0.11.0. This is the code I am using:

```python
pipe_prophet = ForecastingPipeline(steps=[
        ("selection", FeatureSelection(method = 'columns')),
        ("logX", OptionalPassthrough(LogTransformer())),
        ("differencerX", Differencer()),
        ("prophetforecaster", TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("differencer", Differencer()),
            ("prophet", Prophet(yearly_seasonality='auto'))
        ]))  
])

cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = list(range(1,forecasting_horizon +1,1)))

param_grid = [{"logX__passthrough": [False],
               "prophetforecaster__log__passthrough": [False],
               "selection__columns": [[prod_group + ' ADJ'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN','MONTHLY_ROLLING_MIN','MONTHLY_ROLLING_MAX']],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative'],
               "prophetforecaster__prophet__add_country_holidays": [None, {'country_name': 'Italy'}]},
              {"logX__passthrough": [True],
               "prophetforecaster__log__passthrough": [True],
               "selection__columns": [[prod_group + ' ADJ'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN'],[prod_group + ' ADJ','MONTHLY_ROLLING_MEAN','MONTHLY_ROLLING_MIN','MONTHLY_ROLLING_MAX']],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative'],
               "prophetforecaster__prophet__add_country_holidays": [None, {'country_name': 'Italy'}]}]

gscv_prophet = ForecastingGridSearchCV(
            forecaster = pipe_prophet,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )
gscv_prophet.fit(train_model, X = df_ex_train, fh = fh)
```

**Comment by fkiraly** [comment]
oh no.

My suspicion is, it's probably related to this bug/issue here:
https://github.com/alan-turing-institute/sktime/issues/2265

since the `TransformedTargetForecaster` calls the inverse of `Differencer`.

Looks like a bug, so I'll move it over to our issue tracker.

**Reply by fkiraly** [reply]
moved here: https://github.com/alan-turing-institute/sktime/issues/2452

**Comment by fkiraly** [comment]
On another note, did you know that you can specify the same pipeline as
```python
pipe_prophet = ForecastingPipeline(steps=[
       FeatureSelection(method = 'columns') * OptionalPassthrough(LogTransformer()) * Differencer(), 
       OptionalPassthrough(LogTransformer()) * Differencer() * Prophet(yearly_seasonality='auto')
       ]  
)
```
(may require 0.11.2 to not break, and the bug to be fixed, obviously)

I was also thinking about adding more dunders, so the `**` symbol means "apply to `X`" and the `-` symbol means "optional", so:
```python
pipe_prophet = FeatureSelection(method = 'columns') * -LogTransformer() * Differencer()
       ** -LogTransformer() * Differencer() * Prophet(yearly_seasonality='auto')
```

what do you think?

**Reply by anthonygiorgio97** [reply]
ok thanks, but I think we have to find a way to specify which transformation will be applied to X and which to y. I have also another problem with the last release: by using the 0.11.2 version, with the same code (but without Differencer()) I have the following error (here I have used knn instead of prophet, but the pipe is almost the same and the problem is also on prophet and other forecasters):

estimators passed to ForecastingPipeline must be either transformer or forecaster

```
TypeError                                 Traceback (most recent call last)
<[command-3391575316838801]()> in <module>
----> 1 pipe_knn = ForecastingPipeline(steps=[
      2         ("boxcoxX", OptionalPassthrough(BoxCoxTransformer())),
      3         ("detrenderX", OptionalPassthrough(Detrender(forecaster = AutoETS()))),
      4         ("knnforecaster", TransformedTargetForecaster(steps = [
      5             ("boxcox", OptionalPassthrough(BoxCoxTransformer())),

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/compose/_pipeline.py in __init__(self, steps)
    280     def __init__(self, steps):
    281         self.steps = steps
--> 282         self.steps_ = self._check_steps(steps, allow_postproc=False)
    283         super(ForecastingPipeline, self).__init__()
    284         tags_to_clone = [

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/compose/_pipeline.py in _check_steps(self, estimators, allow_postproc)
     61         scitypes = self._get_pipeline_scitypes(estimator_tuples)
     62         if not set(scitypes).issubset(["forecaster", "transformer"]):
---> 63             raise TypeError(
     64                 f"estimators passed to {type(self).__name__} "
     65                 f"must be either transformer or forecaster"

TypeError: estimators passed to ForecastingPipeline must be either transformer or forecaster
```

However the same code works with 0.11.0

**Reply by fkiraly** [reply]
> ok thanks, but I think we have to find a way to specify which transformation will be applied to X and which to y

The transformers before the `**` (double star) are the chain applied to `X`, the transformers after it are those applied to `y`, in that syntax.

**Reply by fkiraly** [reply]
reproduced the problem, my code:

```python
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.boxcox import LogTransformer
from sktime.forecasting.fbprophet import Prophet
from sktime.transformations.series.detrend import Detrender
from sktime.forecasting.ets import AutoETS

pipe_prophet = ForecastingPipeline(steps=[
        ("logX", OptionalPassthrough(LogTransformer())),
        ("detrenderX", OptionalPassthrough(Detrender(forecaster = AutoETS()))),
        ("prophetforecaster", TransformedTargetForecaster(steps = [
            ("log", OptionalPassthrough(LogTransformer())),
            ("prophet", Prophet(yearly_seasonality='auto'))
        ]))  
])
```

**Reply by fkiraly** [reply]
bug report here: https://github.com/alan-turing-institute/sktime/issues/2455

**Reply by fkiraly** [reply]
found the source - type checks were introduced, and the pipelines are an edge case since they are type polymorphic and can have multiple types. Constructor check assumed elements can have only one type, so you get a problem if you nest pipelines. Should be fixed in next release. I´ll also add a test for this to avoid regression of the functionality.

**Reply by anthonygiorgio97** [reply]
Ok this is a code example for my kind of problem (only in 0.11.2; while in 0.11.0 the code works):

```
from sktime.transformations.series.boxcox import LogTransformer, BoxCoxTransformer
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.transformations.series.feature_selection import FeatureSelection
from sktime.forecasting.fbprophet import Prophet
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.datasets import load_longley
y, X = load_longley()

#split train/test both y and X
fh = [1,2,3]
train_model, test_model = temporal_train_test_split(y, fh = fh)
X_train = X[X.index.isin(train_model.index)]
X_test = X[X.index.isin(test_model.index)]

#prophet pipeline
pipe_prophet = ForecastingPipeline(steps=[
        ("selection", FeatureSelection(method = 'columns')),
        ("boxcoxX", OptionalPassthrough(BoxCoxTransformer())),
        ("prophetforecaster", TransformedTargetForecaster(steps = [
            ("boxcox", OptionalPassthrough(BoxCoxTransformer())),
            ("prophet", Prophet(yearly_seasonality='auto'))
        ]))  
])

#cv setup
N_cv_fold = 2
step_cv = 3
cv = ExpandingWindowSplitter(initial_window = len(train_model) - (N_cv_fold - 1) * step_cv - len(fh),
                            start_with_window = True,
                            step_length = step_cv,
                            fh = fh)

param_grid = [{"boxcoxX__passthrough": [False],
               "prophetforecaster__boxcox__passthrough": [False],
               "selection__columns": [['GNPDEFL'],['GNPDEFL', 'GNP']],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative'],
               "prophetforecaster__prophet__add_country_holidays": [None, {'country_name': 'Italy'}]},
              {"boxcoxX__passthrough": [True],
               "prophetforecaster__boxcox__passthrough": [True],
               "selection__columns": [['GNPDEFL'],['GNPDEFL', 'GNP']],
               "prophetforecaster__prophet__seasonality_mode": ['additive','multiplicative'],
               "prophetforecaster__prophet__add_country_holidays": [None, {'country_name': 'Italy'}]}]

#grid search
gscv_prophet = ForecastingGridSearchCV(
            forecaster = pipe_prophet,
            cv=cv,
            param_grid=param_grid,
            verbose = 1
        )

#fit
gscv_prophet.fit(train_model, X = X_train, fh = [1,2,3])
```

Output:

```
TypeError                                 Traceback (most recent call last)
<[command-1608141382505581]()> in <module>
     15 
     16 #prophet pipeline
---> 17 pipe_prophet = ForecastingPipeline(steps=[
     18         ("selection", FeatureSelection(method = 'columns')),
     19         ("boxcoxX", OptionalPassthrough(BoxCoxTransformer())),

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/compose/_pipeline.py in __init__(self, steps)
    280     def __init__(self, steps):
    281         self.steps = steps
--> 282         self.steps_ = self._check_steps(steps, allow_postproc=False)
    283         super(ForecastingPipeline, self).__init__()
    284         tags_to_clone = [

/databricks/python/lib/python3.8/site-packages/sktime/forecasting/compose/_pipeline.py in _check_steps(self, estimators, allow_postproc)
     61         scitypes = self._get_pipeline_scitypes(estimator_tuples)
     62         if not set(scitypes).issubset(["forecaster", "transformer"]):
---> 63             raise TypeError(
     64                 f"estimators passed to {type(self).__name__} "
     65                 f"must be either transformer or forecaster"

TypeError: estimators passed to ForecastingPipeline must be either transformer or forecaster
```

**Reply by anthonygiorgio97** [reply]
> The transformers before the ** (double star) are the chain applied to X, the transformers after it are those applied to y, in that syntax.

So it must be 

```
pipe_prophet = FeatureSelection(method = 'columns') ** -LogTransformer() ** Differencer()
       ** -LogTransformer() * Differencer() * Prophet(yearly_seasonality='auto')
```
because the FeatureSelection() and the first LogTransformer() and Differencer() are for X and not for y

**Reply by fkiraly** [reply]
I think I fixed the "nested pipeline" issue:
https://github.com/alan-turing-institute/sktime/pull/2456

Introducing supposedly helpful typing, and then not thinking about polymorphic edge cases was the problem...

**Reply by fkiraly** [reply]
>> The transformers before the ** (double star) are the chain applied to X, the transformers after it are those applied to y, in that syntax.

> So it must be 

```
pipe_prophet = FeatureSelection(method = 'columns') ** -LogTransformer() ** Differencer()
       ** -LogTransformer() * Differencer() * Prophet(yearly_seasonality='auto')
```
> because the FeatureSelection() and the first LogTransformer() and Differencer() are for X and not for y

My intuition was that the `**` "eats" anything to the left of it, and applies it to `X`.
Because "transformer times transformer is a transformer".
Do you find it more intuitive with only `**`?

I think both ways can be made to parse to the intended effect - is that true?
Whether both are possible would depend on which sequence the dunders are parsed in. Left to right, and first left then right if `NotImplemented`, right?

**Reply by anthonygiorgio97** [reply]
> My intuition was that the ** "eats" anything to the left of it, and applies it to X.
> Do you find it more intuitive with only **?

Ok clear. Yes I find more intuitive with only ** each time. But it's the same

**Reply by fkiraly** [reply]
so, overall - good or bad idea, in your opinion?

**Reply by anthonygiorgio97** [reply]
Good idea for me

**Comment by fkiraly** [comment]
also looks like you have a hierarchical problem? (product group etc)

you may also want to try out the `Reconciler`, see here:
https://github.com/sktime/sktime-tutorial-pydata-berlin-2022/blob/main/notebooks/3_hierarchical_global_forecasting.ipynb
(needs 0.11.2)

Unrelated to your original problem, but I thought you may like to know.

**Comment by fkiraly** [comment]
So, @anthonygiorgio97, we've now released 0.11.3, this should fix two of the issues above:

* the bug with nesting pipelines is fixed https://github.com/alan-turing-institute/sktime/pull/2456
* `Differencer` now has a new `na_handling` parameter https://github.com/alan-turing-institute/sktime/pull/2487, if you set it to `fill_zero` it should be compatible with pipelines that have elements which require indices to be unchanged

The underlying issue with `Prophet` is probably this one, it is coming directly from facebook `prophet`:
https://github.com/alan-turing-institute/sktime/issues/2564
(does this sound plausible?)
Someone might be adding a workaround in the future.

**Reply by anthonygiorgio97** [reply]
Thanks for all. Now my code works again with 0.11.3.

I was thinking that another workaround to not give a wrong value (in this case 0) at the beginning of the differenced series in the pipeline is to save in a variable the first difference, then use `Differencer` from the second value of the time series with the option `"keep_na"` and then use `Imputer` the fill the first na value with the first difference

For example, given a time series like `[a,b,c,d]` the algorithm will be:

1. save first_diff <- b-a

2. drop `a` from `[a,b,c,d]` -> `[b,c,d]`

3. create a pipeline for `[b,c,d]`:

      -use `Differencer` with `"keep_na"` : `[?, c-b, d-c]`

      -use `Imputer` with `method = "costant"` and `arg = first_diff`  : `[b-a, c-b, d-c]`

      -apply forecaster

This is more robust then apply the pipeline to `[0,b-a,c-b,d-c]` which is wrong because it came from the time series `[a,a,b,c,d]` I think, or maybe I'm exaggerating this issue which it is not so important, I don't know.

**Reply by fkiraly** [reply]
huh, that´s a good point. I just checked, and the differencer was not remembering differences it had previously computed, or data it had previously seen. That should simply be filled in, if it is available.

Is it that what you mean?
(if not, would appreciate another explanation, with three paragraphs: what is the situation, what is current behaviour in that situation, and what is proposed/desired behaviour)

**Reply by fkiraly** [reply]
?
